orca-core/src/test/groovy/com/netflix/spinnaker/orca/pipeline/util/RegionCollectorSpec.groovy

```diff
@@ -0,0 +1,107 @@
+package com.netflix.spinnaker.orca.pipeline.util
+
+import spock.lang.Specification
+
+import static com.netflix.spinnaker.orca.test.model.ExecutionBuilder.pipeline
+import static com.netflix.spinnaker.orca.test.model.ExecutionBuilder.stage
+
+class RegionCollectorSpec extends Specification {
+  RegionCollector regionCollector
+
+  def setup() {
+    regionCollector = new RegionCollector()
+  }
+
+  def "should find all regions from subsequent deploys"() {
+    when:
+    def regionSet = regionCollector.getRegionsFromDescendentStages(execution.stageByRef("1"))
+
+    then:
+    regionSet == allDeployRegions
+
+    where:
+    deployClusters = [
+      [
+        cloudProvider: "aws",
+        availabilityZones: [ region1: [], region2: []]
+      ],
+      [
+        cloudProvider: "aws",
+        availabilityZones: [ region3: [] ]
+      ]
+    ]
+
+    allDeployRegions = deployClusters.sum {it->it.availabilityZones.keySet()}
+
+    execution = pipeline {
+      stage {
+        name = "findImage"
+        type = "findImage"
+        refId = "1"
+        context = [ cloudProvider: "aws"]
+      }
+
+      stage {
+        name = "deploy"
+        type = "deploy"
+        context = [
+          clusters: deployClusters
+        ]
+        requisiteStageRefIds = ["1"]
+      }
+    }
+  }
+
+  def "should find all regions from subsequent canary deploys"() {
+    when:
+    def regionSet = regionCollector.getRegionsFromDescendentStages(execution.stageByRef("1"))
+
+    then:
+    regionSet == allDeployRegions
+
+    where:
+    deployCluster1 = [
+      cloudProvider: "aws",
+      availabilityZones: [ region1: [], region2: []]
+    ]
+
+    deployCluster2 = [
+      cloudProvider: "aws",
+      availabilityZones: [ region3: [] ]
+    ]
+
+    deployCluster3 = [
+      cloudProvider: "aws",
+      availabilityZones: [ region4: [] ]
+    ]
+
+    allDeployRegions = [deployCluster1, deployCluster2, deployCluster3].sum {it -> it.availabilityZones.keySet()}
+
+    execution = pipeline {
+      stage {
+        name = "findImage"
+        type = "findImage"
+        refId = "1"
+        context = [ cloudProvider: "aws"]
+      }
+
+      stage {
+        name = "canary"
+        type = "canary"
+        context = [
+          clusterPairs: [ [
+              baseline: deployCluster1,
```

Comment:
test could be significantly shortened if you had a helper method create the clusters ...

```
baseline: cluster(["us-west-2", "us-east-1"])
```

For a test like this, I think it's fine to use actual region names and then just reference them explicitly in your `then:` (vs. needing to compare against a variable)
...

```
regionSet == ["us-west-2", "us-east-1", "eu-west-1"] // this would be fine imo
```

Reply:
yeah, i think that makes sense. fixup